File: prisma/refs/java/word-ladder.java
import java.util.*;

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> words = new HashSet<>(wordList);
        if (!words.contains(endWord)) return 0;
        Set<String> frontier = new HashSet<>(Collections.singletonList(beginWord));
        int steps = 1;
        while (!frontier.isEmpty()) {
            Set<String> next = new HashSet<>();
            for (String w : frontier) {
                if (w.equals(endWord)) return steps;
                char[] chars = w.toCharArray();
                for (int i = 0; i < chars.length; i++) {
                    char orig = chars[i];
                    for (char c = 'a'; c <= 'z'; c++) {
                        chars[i] = c;
                        String cand = new String(chars);
                        if (words.remove(cand)) next.add(cand);
                    }
                    chars[i] = orig;
                }
            }
            frontier = next;
            steps++;
        }
        return 0;
    }
}

File: prisma/refs/py/word-ladder.py
import string


def ladder_length(begin_word, end_word, word_list):
    words = set(word_list)
    if end_word not in words:
        return 0
    frontier, steps = {begin_word}, 1
    while frontier:
        nxt = set()
        for w in frontier:
            if w == end_word:
                return steps
            for i in range(len(w)):
                for c in string.ascii_lowercase:
                    cand = w[:i] + c + w[i + 1:]
                    if cand in words:
                        words.discard(cand)
                        nxt.add(cand)
        frontier = nxt
        steps += 1
    return 0
